etc/docs/Design.md

```diff
@@ -0,0 +1,100 @@
+# Things to cover
+
+## API
+
+### High level Library API
+
+The library API should provide the following methods to the clients:
+
+* Initialize an executor instance with the following settings
+  * Limits for resources
+    
+
+* Execute a command with the supplied arguments, that returns a pointer to a `Command` object that can be later used to manipulate the underlying process
+
+### GRPC API definition
+
+The GRPC API defines the following methods:
+
+* ExecuteCommand: Starts a command with limited resources in its own namespace with the supplied arguments
+* StopExecution : Stop a certain execution 
+* ListExecutions: List all the executions the connected client has rights to
+* GetExecution: Gets metadata about a certain execution defined by an execution ID
+* GetOutStream: Get a stream of lines of standard output (stdout & stderr) of a certain execution
+* GetOut: Get all lines of standard output (stdout & stderr) that has been written so far by the process
+
+To read about the exact details of the API you can look at [the proto file](api.proto)
+
+### Streaming data
+
+In order to be able to stream and also send the full output of the processes we need to store the output of this in memory for simplicty, in order to be able to keep memory from being overloaded a simple recylcing of these processes will be implemented. See [Tradeoffs](#tradeoffs) section for information about why this solution should not be used in production  
+
+## Security
+
+### Cipher suite
+
+In order to be secure the app will use the TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 cypher suite.
+
+It suits our needs to be a strong cypher suite by using AES, SHA-2 and ephemeral Diffie-Hellman key exchange, even though the ephemeral key exchange thakes more compute power, with the amount of the frequency of the connections (processes are started, then waited on, and output read) the added security is worth it
+
+### Authentication
+
+### mTLS
+
+For the mTLS the following certificates will be generated by a generate-certs shell script to enable easy testing of the application (see the Tradeoff section for more details):
+
+* A root CA certificate to sign client and server CSR-s 
+* A x509 certificate for the server signed by the root CA
+* Two x509 certificates one for each of two clients, that could be used to connect to the applicaiton
+
+The root CA will use a private key that's 4096 bits and the client and server certs will use private keys with 2048 bits
+
+### Authorization
+
+To keep things simple, the following statement should hold for every process started on the server:
+
+`Only clients authenticated with the client certs hacing the same CN can get any information about a certain process`
```

Comment:
This sounds good, but just to clarify - "can get any information" - does that also include stopping jobs?

Reply:
Yeah by "can get any information" means that it will be able to do absolutely nothing about the execution, nor execute any action on it. I will change the wording to make it more clear